mono/arch/arm/arm-codegen.h

```diff
@@ -21,7 +21,7 @@ typedef unsigned int armword_t;
 	void __inline _arm_emit(arminstr_t** p, arminstr_t i) {**p = i; (*p)++;}
 #	define ARM_EMIT(p, i) _arm_emit((arminstr_t**)&p, (arminstr_t)(i))
 #else
-#	define ARM_EMIT(p, i) do { arminstr_t *__ainstrp = (void*)(p); *__ainstrp = (arminstr_t)(i); (p) = (void*)(__ainstrp+1);} while (0)
+#	define ARM_EMIT(p, i) do { *(arminstr_t*)(p) = (arminstr_t)(i); *(void**)&(p) = (arminstr_t*)(p) + 1; } while (0)
```

Comment:
This line bothers me. The address-taken of `p` will likely impede the optimizer.
Some ideas.

1) same thing but under `ifdef __cplusplus`. Ok short term but still bad long term.

2) change all `p` to be the same type; maybe too big/unncecessary of a change.

3) assert that `sizeof(*p) == 1 or 2 or 4` (or `% sizeof(arminstr) == 0`), and increment `p` accordingly, with constant division even. If option 3 at all nerve-wracking, can be under `ifdef cplusplus`.

Option 3 is wearing well in my head.

Reply:
I am working on part 3 in a separate branch  PR. Unless reviewers don't mind the address-taken as much as I do..

Reply:
If the address-taken is a bother, as it is to me, let's consider https://github.com/mono/mono/pull/9982 for this part. There will be a merge conflict and 9982 wins, if consensus.